WTF: Fix unsigned overflows/underflows in string conversion

LumioseSil opened the pull request:
#### a83ee572364b9ae35c05e5eb6a36d502515d5c22
<pre>
WTF: Fix unsigned overflows/underflows in string conversion
<a href="https://bugs.webkit.org/show_bug.cgi?id=256164">https://bugs.webkit.org/show_bug.cgi?id=256164</a>

Reviewed by David Kilzer.

Although we check that integer is less than 0 before negating it, the
UBSanitizer keeps tripping up at this section. The proper fix would be
to use std::make_unsigned_t&lt;IntegerType&gt; here, to convert the variable
to its unsigned equivalent.

* Source/WTF/wtf/text/IntegerToStringConversion.h:
  (writeIntegerToBuffer): Cast to unsigned.

Canonical link: <a href="https://commits.webkit.org/263866@main">https://commits.webkit.org/263866@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/713b9e6b8613cca556a3984947de391486b11db0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5037 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5167 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5344 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6556 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5128 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5359 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5142 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5380 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5123 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5208 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4518 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6573 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2695 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4516 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/9510 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/4172 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4558 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4589 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6194 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/4649 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5003 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4100 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/5130 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4489 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1293 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8566 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/5274 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/697 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4855 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1408 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WTF/wtf/text/IntegerToStringConversion.h:
```diff
@@ -91,8 +91,8 @@ inline void writeIntegerToBuffer(IntegerType integer, CharacterType* destination
         if (integer < 0)
-            return writeIntegerToBufferImpl<CharacterType, typename std::make_unsigned_t<IntegerType>, NegativeNumber>(-integer, destination);
-        return writeIntegerToBufferImpl<CharacterType, typename std::make_unsigned_t<IntegerType>, PositiveNumber>(integer, destination);
+            return writeIntegerToBufferImpl<CharacterType, typename std::make_unsigned_t<IntegerType>, NegativeNumber>(std::make_unsigned_t<IntegerType>(-integer), destination);
```
As far as I can tell, the syntax here is incorrect because `std::make_unsigned_t<IntegerType>` describes a _type_, not a conversion function.

The proper way to do this is to use `static_cast<>()` with the type like this:

```
static_cast<std::make_unsigned_t<IntegerType>>(-integer)
```

All four call sites need to be fixed.

ddkilzer replied:
Oops, I forgot that `type(expr)` is also valid in C++:
<https://en.cppreference.com/w/cpp/language/explicit_cast>

This is fine.

ddkilzer reviewed (changes requested):
Please update the PR to use `static_cast<>()` with `std::make_unsigned_t<>`.

ddkilzer reviewed (approved):
r=me

LumioseSil committed:
WTF: Fix unsigned overflows/underflows in string conversion
https://bugs.webkit.org/show_bug.cgi?id=256164

Reviewed by David Kilzer.

Although we check that integer is less than 0 before negating it, the
UBSanitizer keeps tripping up at this section. The proper fix would be
to use std::make_unsigned_t<IntegerType> here, to convert the variable
to its unsigned equivalent.

* Source/WTF/wtf/text/IntegerToStringConversion.h:
  (writeIntegerToBuffer): Cast to unsigned.

Canonical link: https://commits.webkit.org/263866@main
